End to end prototype

KinanBab opened the pull request:
End to end prototype with just enough features to handle the following sample P4 program:
https://github.com/googleinterns/p4-symbolic/blob/end2end/p4-samples/port-table/table.p4

The workflow proceeds as follows:
1. p4c is run on the p4 program to produce bmv2 json and p4info (use `` to do this)
1. Bmv2 JSON is parsed into a primitive protobuf that matches the bmv2 exactly.
2. p4info is parsed with pdpi
3. bmv2 protobuf, pdpi, and table entries are combined and transformed into our IR representation
4. our IR representation is fed to the Analyzer class, which traverses the program, creating a bunch of symbolic variables and constraints on them guided by the IR.
5. The analyzer class can be queried by giving it a table name and a row in that table, and it will generate a packet that matches that row. This can be repeated many times.

The prototype can be run using:
```bash
# from repo root
# produce .json and p4info files
cd p4-samples/port-table && make build
# compile our project
bazel build //p4_symbolic:main
# run the binary on the sample program
./bazel-bin/p4_symbolic/main --p4info=p4-samples/port-table/table.pb.txt --bmv2=p4-samples/port-table/table.json --entries=p4-samples/port-table/entries.txt
```

The binary will output a debugging smt program, followed by the testing packets.

Remaining tasks to complete this PR: 
- [x] Refactor p4_symbolic/main.cc into an IR testing file and a general binary.
- [x] Extract packet in a better format from model and write to a file.
- [x] Dump debugging smt program to a file, if flagged.
- [x] Minor code clean ups (AnalyzeTable function is too big).
- [x] Add generic functions for iterating over tables and rows, and use them in the binary, instead of using hardcoded values.

Shortcuts and unsupported features are marked clearly with TODOs in the respective location of the code. These will be addressed in future PRs.

KinanBab committed:
Add symbolic analyzer stub

KinanBab committed:
Clean up duplicate in IR

KinanBab committed:
clean up unneeded files

KinanBab committed:
begin working on symbolic z3 interface

KinanBab committed:
Initial end-2-end prototype

TODO:
- Refactor p4_symbolic/main.cc into an IR testing file and a binary
- Extract packet in a better format from model
- Review a better way of doing the StatusOr<z3::expr> without pointer?
- Review output smt lib program, improve how it is printed and logged
- Clean up the AnalyzeTable function, maybe split it into several smaller functions?
- Add functions for iterating over tables, rows, cross-products etc

KinanBab committed:
Split main.cc functionality and IR testing functionality into two files, add parser.cc which is used by both files for parsing

KinanBab commented:
Test refactoring is done, but I pushed it to a separate branch 'refactor-testing' to keep this PR small.

KinanBab committed:
Clean up code and add additional integrity constraints to ensure proper flow of match to action

KinanBab committed:
Extract packet from model properly

KinanBab committed:
Install z3 in ci docker

KinanBab committed:
Clean up Z3 internal memory, valgrind shows no more memory errors

KinanBab committed:
Dump debug smt formula to file when --debug is provided

KinanBab committed:
Add golden file testing for the main binary, including debugging smt formula

Do not use insert_or_assign because of bazel weird issue with c++17

KinanBab committed:
Improve packet extraction
1. main.cc loops over all tables and rows
2. symbolic.cc returns packets as an (ordered) map, so they are always printed consistently.

konne88 commented on p4_symbolic/main.cc:
```diff
@@ -106,7 +108,39 @@ int main(int argc, char *argv[]) {
-  std::cout << ir_status.value().DebugString() << std::endl;
+  // Analyze program symbolically.
+  p4_symbolic::symbolic::Analyzer analyzer;
+  p4_symbolic::ir::P4Program program = ir_status.value();
```
You should use RETURN_IF_ERROR/ASSIGN_OR_RETURN/RET_CHECK macros for error handling, so avoid distracting from the main code. Here, I would recommend pulling all the code into a Main function that returns a Status, and then do the std::cerr << status; return 1; only once.

konne88 replied:
Why did you mark this resolved? It doesn't look changed to me

konne88 commented on p4_symbolic/symbolic/symbolic.h:
```diff
@@ -0,0 +1,114 @@
+#include "absl/status/status.h"
+#include "p4_pdpi/utils/status_utils.h"
+#include "p4_symbolic/ir/ir.pb.h"
+#include "z3++.h"  // added as a system dependency for now.
```
can you put

TODO(your google username): added as ...

konne88 commented on p4_symbolic/symbolic/symbolic.h:
```diff
@@ -0,0 +1,114 @@
+
+  // Symbolic constraints that describe relationships between various variables
+  // and fields in the program.
+  std::vector<z3::expr> kConstraints;
```
Are these constants? If so, define them const, if not, they should be lowercase and end with _

KinanBab replied:
yes this is my bad, I misread the style guide

konne88 commented on p4_symbolic/symbolic/symbolic.h:
```diff
@@ -0,0 +1,114 @@
+  Analyzer& operator=(Analyzer&&) = delete;
+
+  // Entry point
+  absl::Status Analyze(ir::P4Program);
```
Please use explicit dataflow. It's very hard to tell what is going on here, when every function you are calling can mutate the entire state. I propose you remove this class, and have functions that take and return only the data they need.

konne88 commented on p4_symbolic/symbolic/symbolic.h:
```diff
@@ -0,0 +1,114 @@
+
+  // API for finding test packets after analysis.
+  pdpi::StatusOr<std::map<std::string, std::string>> FindPacketHittingRow(
+      const std::string&, int);
```
give all function parameters a name, everywhere

konne88 commented on p4_symbolic/symbolic/symbolic.h:
```diff
@@ -0,0 +1,114 @@
+
+  // Symbolic constraints that describe relationships between various variables
+  // and fields in the program.
+  std::vector<z3::expr> kConstraints;
```
Instead of creating constraints, I think we should write the code to create symbolic expressions, and only at the very end do we assert something. In particular, the entire switch pipeline should be a symbolic expression. The expression should have the access to the input packet, and should produce a trace with all the information about how the packet was forwarded (which table entry was hit, was the packet dropped, etc). You can look at the code that I shared with you for more detail. The user can then write an assertion over that expression, e.g. packet must not have dropped, and must have hit table entry FOO. 

If you follow that style, your program will look very nice, because it looks just like a non-symbolic P4 simulator that happens to operate with symbolic values and operations.

KinanBab committed:
Update ir and bmv2 test files with new API

KinanBab committed:
Split main function in main binary and tests to a helper returning absl::Status and only do error handling in main.

KinanBab committed:
Clean up

* Define an explicit type for an output packet.
* Clean up class member names.
* Add parameter names to header.

KinanBab committed:
Use Patched StatusOr that supports non-trivially constructable value types

KinanBab committed:
Re-organize the analysis code as well as the resulting SMT program.

Both now look more like a regular interpretation of the underlying P4
program, just done symbolically.

The analysis code maintains no mutable state, except for the scope of an
action body.

State is organized as helper structs. Analysis code is organized as
functions, each matching a unique construct from our IR. These functions
return a symbolic expression, representing the symbolic interpretation
of the corresponding IR construct, or an error message if invalid
constructs are encountered.

The SMT program consists of a single constraint per table. The
constraint is a conjunction of alias definitions (one per entry), and a
single if-elseif-else construct, matching on every entry and ending with
a special 'dropped' alias. The body of every match is a symbolic
expression constraining the parameters of the corresponding action and
asserting that action's body symbolically.

konne88 commented on p4_symbolic/symbolic/symbolic.h:
```diff
@@ -0,0 +1,56 @@
+
+// Symbolically Analyzes the given program, generating z3 symbolic expressions
+// and that are later solved to produce concerete packets.
+pdpi::StatusOr<SolverState> AnalyzeProgram(const ir::P4Program &program);
```
symbolic_ingress_port = FreePort();
symbolic_ingress_header = FreeHeader();
symbolic_trace = RunP4PipelineSymbolically(p4_program, table_entries, symbolic_ingress_port, symbolic_ingress_header))

konne88 commented on p4_symbolic/symbolic/structs.h:
```diff
@@ -0,0 +1,101 @@
+  // expressions must be created using the exact same context.
+  // Using a smart pointer address these issues and automatically manages the
+  // memory.
+  std::unique_ptr<z3::context> context;
```
Don't use this inside of P4 pipeline, create all variables/assertions outside.

konne88 commented on p4_symbolic/symbolic/table.h:
```diff
@@ -0,0 +1,40 @@
+namespace symbolic {
+namespace table {
+
+pdpi::StatusOr<z3::expr> AnalyzeTable(const ir::Table &table,
```
trace = RunTable(p4_table, table_entries, state)

trace consists of new state, and then table entry that was hit to give to the assertion

konne88 commented on p4_symbolic/symbolic/action.h:
```diff
@@ -0,0 +1,52 @@
+
+// Symbolically Analyzes the given program, generating z3 symbolic
+// expressions and that are later solved to produce concerete packets.
+pdpi::StatusOr<ActionSymbolicTrace> AnalyzeAction(
```
state = RunAction(action, state)

konne88 replied:
Rerun action analysis every time.

konne88 commented on p4_symbolic/symbolic/names.h:
```diff
@@ -0,0 +1,60 @@
+// Copyright 2020 Google LLC
```
This will go away

konne88 commented on p4_symbolic/symbolic/structs.h:
```diff
@@ -0,0 +1,101 @@
+// Copyright 2020 Google LLC
```
Put these with their appropriate functions

KinanBab committed:
Initial restructuring

* Copy header files from headers PR
* Modify header files slightly: SymbolicContext is returned as a pointer to avoid std::move issue, z3 context has to be saved to avoid seg faults
* Implement functions to create free symbolic context, extract concrete context from it
* Implement symbolic.cc with empty table.cc and action.cc
* Update main.cc to use new format, certain unsupported features are left as comments

KinanBab committed:
Update TableEntry format: TableEntries are now seperate from IR

KinanBab committed:
Continued restructuring

* Made additional needed changes to signatures and headers to pass context around.
* Implemented cc body code for evaluating tables and actions
* Improved util functions for defining unconstrained/free symbolic headers and metadata.
* Add util for merging two symbolic states with a condition.
* Improve debugging SMT program display

KinanBab committed:
Make z3::context a global variable managed by our library

KinanBab committed:
Better management of global Z3_CONTEXT and singleton SolverState

konne88 commented on p4_symbolic/symbolic/symbolic.cc:
```diff
@@ -0,0 +1,126 @@
+namespace p4_symbolic {
+namespace symbolic {
+
+z3::context *Z3_CONTEXT = nullptr;
```
Can we just initialize this variable once, and keep it forever? How the code initializes and frees it right now, is hard to follow for me.

konne88 commented on p4_symbolic/symbolic/symbolic.cc:
```diff
@@ -0,0 +1,126 @@
+  return solver_state;
+}
+
+pdpi::StatusOr<ConcreteContext> Solve(SolverState *solver_state,
```
Can we change this to optional? It's not really an error if the solver doesn't find a solution.

KinanBab replied:
Makes sense

konne88 commented on p4_symbolic/symbolic/symbolic.cc:
```diff
@@ -0,0 +1,126 @@
+
+z3::context *Z3_CONTEXT = nullptr;
+
+pdpi::StatusOr<SolverState *> EvaluateP4Pipeline(
```
Is there are reason to return the pointer? Why not just return the value?

KinanBab replied:
Yes, the SolverState has a z3::context and z3::solver in it. These do not have copy constructors. pdpi::StatusOr<> requires a copy constructor. Also, because I wanted SolverState to be essentially a singleton.

konne88 commented on p4_symbolic/symbolic/symbolic.cc:
```diff
@@ -0,0 +1,126 @@
+                        "been destructed!");
+  }
+
+  // Context to use for defining z3 variables, etc..
```
Instead of initializing the context here, can we just initialize it once for the whole program. There is a lot of subtly around global variables that have destructors (go/totw/110), but the TLDR is:

z3::context& Z3Context() {
  static z3::context* context = new z3::context();
  return *context;
}

KinanBab replied:
Actually, there is no need for Z3_CONTEXT to even be a pointer.

konne88 replied:
So if you have the Z3Context function, you just always access the context through that function. Because the context inside the function is static, it's essentially a global variable. Inside the function, the context has to be a pointer, otherwise, the object would be destructed when the function returns (I believe).

konne88 replied:
Oh actually, the destructor will be called at the end of the program, but I believe we don't want that, due to some subtle race conditions etc as described in go/totw/110

konne88 commented on p4_symbolic/symbolic/symbolic.cc:
```diff
@@ -0,0 +1,126 @@
+  // Context to use for defining z3 variables, etc..
+  Z3_CONTEXT = new z3::context();
+  z3::solver *z3_solver = new z3::solver(*Z3_CONTEXT);
+
```
You should basically never call new. Can you create a unique_ptr instead?

konne88 reviewed (changes requested):
Looks great. My main feedback is that I think the global variable initialization and destruction seems too complex. Can we just initialize the context once on program start?

KinanBab commented:
Makes sense. It is a bit annoying how the context is managed right now. I can make this change to simplify it.

KinanBab committed:
Pointers clean up

* Make global Z3_CONTEXT a variable that is always intialized.
* Augment with a boolean flag to detect if previous calls terminated.
* Do not use new anywhere
* Return SolverState as a unique_ptr to signify it is managed by the calling context.
* Functions take SolverState as a const &
* Solve returns an optional that is empty when z3 fails to solve the constraints (it is not an error).

KinanBab commented:
Your comments where very helpful. I did some more digging, and got rid of all pointers. Z3_CONTEXT is now just a plain value that is always initialized. SolverState is returned as a unique_ptr.

KinanBab committed:
Use go/totw/110 construction to define a safe global variable with respect to intialization and destruction.

konne88 commented on p4_symbolic/symbolic/symbolic.cc:
```diff
@@ -0,0 +1,119 @@
+namespace p4_symbolic {
+namespace symbolic {
+
+z3::context &Z3_CONTEXT() {
```
Wrong naming style for a function. Call it Z3Context.

KinanBab committed:
Merge pull request #30 from googleinterns/refactor-testing

Split IR testing and main binaries, add golden files testing for the new symbolic pipeline.

KinanBab committed:
Fix naming style of global z3::context function

KinanBab committed:
Code styling
